keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,350 @@
+- closed failures which break on old clients (making networkpolicy api evolution very hard)
+- open failures with lots of warnings and docs about how old clients might 
+overinterpret the allowed connectivity for a policy
+
+Both of these are impractical due to the nature of security tools, which need to 
+be robust and explicit.  Although failing in a closed manner is acceptable... it is obviously unfortunate,
+as it would obsolete all network policy providers *not* eagerly adopting such a new semantic, which is
+overly restrictive and against the spirit of community-friendly, safe API evolution which is so important
+to a large project like Kubernetes with many external functional plugins.
+
+Ideally, a fix here would “just work” for “back-rev” implementations of NetworkPolicy 
+(which are out-of-tree - generally implemented by cni plugins) and would not 
+introduce new surprises for users. But in practicality this is not the case 
+because, by adding a new selection data structure, the logical inclusion/exclusion 
+properties of namespace and pod selectors become drastically different, as shown in the table below:
+
+
+In this table "policy-reader" might be something like the "calico-controller" or any other cni agent
+that reads the API Servers network policy portfolio and responds to it over time.
+
+|         input policy         | policy-reader (old)    | client (new)        |   |   |
+|------------------------------|------------------------|---------------------|---|---|
+| nsNameSelector + podSelector | ANY ns, pod restricted | ns + pod restricted |   |   |
+| nsNameSelector               | invalid (empty peer)   | ns restricted       |   |   |
+| podSelector                  | ANY namespace          | ANY namespace       |   |   |
+
+In an older client (one which predated the addition of a new, "nsNameSelector" field, we find
+that there is a scenario (`ANY ns, pod restricted`), wherein an old interpretation of the API
+assumes that the ONLY restricting field is a pod label.  Thus, although the user intended to
+restrict all traffic to a namespace/pod combination, the policy implementer does something **much**
+more promiscuous.
+
+This is all avoidable if we could assume, in all cases, that namespaces have a default label.
+
+There may also be many other usecases where the a security or permissions boundary might be conveniently defined in terms
+of a namepsace name, and that such a definition would be most easily referenced by a default label.
+
+### Goals
+
+- Add an immutable default label to ALL namespaces, as the namespace name so this 
+can be used as selector by arbitrary components.
+
+### Non-Goals
+
+* Publishing name labels on all resources
+* Exposing arbitrary fields as selectors
+
+## Proposal
+
+We propose to label of all namespaces, by default, with a reserved label 
+(i.e. “kubernetes.io/metadata.name” or some such), to allow easy selection.
+
+If this label is missing, it is added by the apiserver
```

Comment:
As part of existing namespace admission controller?  Assume we will have a loop in controller manager that applies it for existing namespaces if not present?

Reply:
@abhiraut ^

Reply:
ignore question about admission, but assume we will want some path to apply label after upgrade.  maybe kube storage version migration is sufficient.

Reply:
Great catch .  Talking with @andrewsykim and @rikatz about this.  We'll add details to  the upgrade/downgrade section and disambiguate this .

Reply:
updated to reflect the new thinking around defaulting APIs